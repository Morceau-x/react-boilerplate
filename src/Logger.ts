import { AppDomain } from './utils/Basics';

const isDebugMode = (): boolean => process.env.NODE_ENV === 'development';

export type LoggerSpecs = {
    log: (domain: AppDomain, ...values: unknown[]) => void;
    warn: (domain: AppDomain, ...values: unknown[]) => void;
    error: (domain: AppDomain, ...values: unknown[]) => void;
    logValue: (domain: AppDomain, ...values: unknown[]) => (value: unknown) => void;
    warnValue: (domain: AppDomain, ...values: unknown[]) => (value: unknown) => void;
    errorValue: (domain: AppDomain, ...values: unknown[]) => (value: unknown) => void;

    logDebug: (domain: AppDomain, ...values: unknown[]) => void;
    warnDebug: (domain: AppDomain, ...values: unknown[]) => void;
    errorDebug: (domain: AppDomain, ...values: unknown[]) => void;
    logValueDebug: (domain: AppDomain, ...values: unknown[]) => (value: unknown) => void;
    warnValueDebug: (domain: AppDomain, ...values: unknown[]) => (value: unknown) => void;
    errorValueDebug: (domain: AppDomain, ...values: unknown[]) => (value: unknown) => void;
};

class ConsoleLogger implements LoggerSpecs {
    log = (domain: AppDomain, ...values: unknown[]): void => console.log(domain, ...values);
    warn = (domain: AppDomain, ...values: unknown[]): void => console.warn(domain, ...values);
    error = (domain: AppDomain, ...values: unknown[]): void => console.error(domain, ...values);
    logValue =
        (domain: AppDomain, ...values: unknown[]): ((value: unknown) => void) =>
        (value: unknown) =>
            console.log(domain, ...values, value);
    warnValue =
        (domain: AppDomain, ...values: unknown[]): ((value: unknown) => void) =>
        (value: unknown) =>
            console.log(domain, ...values, value);
    errorValue =
        (domain: AppDomain, ...values: unknown[]): ((value: unknown) => void) =>
        (value: unknown) =>
            console.log(domain, ...values, value);

    logDebug = (domain: AppDomain, ...values: unknown[]): void => {
        isDebugMode() ? console.log(domain, ...values) : null;
    };
    warnDebug = (domain: AppDomain, ...values: unknown[]): void => {
        isDebugMode() ? console.warn(domain, ...values) : null;
    };
    errorDebug = (domain: AppDomain, ...values: unknown[]): void => {
        isDebugMode() ? console.error(domain, ...values) : null;
    };
    logValueDebug =
        (domain: AppDomain, ...values: unknown[]): ((value: unknown) => void) =>
        (value: unknown) => {
            isDebugMode() ? console.log(domain, ...values, value) : null;
        };
    warnValueDebug =
        (domain: AppDomain, ...values: unknown[]): ((value: unknown) => void) =>
        (value: unknown) => {
            isDebugMode() ? console.warn(domain, ...values, value) : null;
        };
    errorValueDebug =
        (domain: AppDomain, ...values: unknown[]): ((value: unknown) => void) =>
        (value: unknown) => {
            isDebugMode() ? console.error(domain, ...values, value) : null;
        };
}

export const Logger: LoggerSpecs = new ConsoleLogger();
