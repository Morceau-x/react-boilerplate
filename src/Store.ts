import createSagaMiddleware from 'redux-saga';
import { configureStore } from '@reduxjs/toolkit';
import { WindowSlice } from './redux/WindowSlice';
import { ThemeSlice } from './redux/ThemeSlice';
import { all, SagaGenerator } from 'typed-redux-saga';
import { TestSlice } from './redux/Test';

const reducers = {
    window: WindowSlice.slice.reducer,
    theme: ThemeSlice.slice.reducer,
    test: TestSlice.slice.reducer,
};

const testSagas = TestSlice.sagas();

const watchers: SagaGenerator<unknown>[] = [...testSagas];

const sagas = function* root(): Generator {
    yield* all(watchers);
};

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
    reducer: { ...reducers },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware({ thunk: false }).concat(sagaMiddleware),
});

sagaMiddleware.run(sagas);
