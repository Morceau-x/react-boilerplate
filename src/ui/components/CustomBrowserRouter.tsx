import React, { useContext } from 'react';
import { BrowserRouterProps } from 'react-router-dom';
import { createBrowserHistory } from 'history';
import { Router } from 'react-router';

export const History = createBrowserHistory({ window });
export const HistoryProvider = React.createContext(History);

const CustomInternalBrowserRouter: React.FC<BrowserRouterProps> = ({ basename, children }) => {
    const history = useContext(HistoryProvider);

    const [state, setState] = React.useState({
        action: history.action,
        location: history.location,
    });

    React.useLayoutEffect(() => history.listen(setState), [history]);

    return (
        <Router basename={basename} location={state.location} navigationType={state.action} navigator={history}>
            {children}
        </Router>
    );
};

export const CustomBrowserRouter: React.FC<BrowserRouterProps> = (props) => {
    return (
        <HistoryProvider.Provider value={History}>
            <CustomInternalBrowserRouter {...props} />
        </HistoryProvider.Provider>
    );
};
