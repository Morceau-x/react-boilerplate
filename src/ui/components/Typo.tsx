import React, { PropsWithChildren } from 'react';
import { css, cx } from '@emotion/css';
import { Theme } from '../../models/ThemeModel';
import { useStyles } from '../../hooks/useTheme';

type TextType =
    | 'Title1'
    | 'Title2'
    | 'Title3'
    | 'Title4'
    | 'Paragraph1'
    | 'Paragraph2'
    | 'Paragraph3'
    | 'Paragraph4'
    | 'Paragraph5'
    | 'Error'
    | 'Warning'
    | 'ErrorMessage'
    | 'SuccessMessage'
    | 'ValidationError';

type TypoProps = {
    textType: TextType;
    textStyle?: string;
    Component?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p' | 'span';
};

export const Typo: React.FC<PropsWithChildren<TypoProps>> = ({ textType, textStyle, children, Component = 'p' }) => {
    const styles = useStyles(makeStyles, textType);
    return <Component className={cx(styles.text, textStyle)}>{children}</Component>;
};

const makeStyles = (theme: Theme, textType: TextType) => ({
    text: (() => {
        switch (textType) {
            case 'Title1':
                return css`
                    color: ${theme.colors.typoMax};
                    font-size: ${theme.fontSizes.XXL}rem;
                    font-weight: bold;
                `;
            case 'Title2':
                return css``;
            case 'Title3':
                return css``;
            case 'Title4':
                return css``;
            case 'Paragraph1':
                return css`
                    color: ${theme.colors.typoMedium};
                    font-size: ${theme.fontSizes.M}rem;
                    font-weight: normal;
                `;
            case 'Paragraph2':
                return css``;
            case 'Paragraph3':
                return css``;
            case 'Paragraph4':
                return css``;
            case 'Paragraph5':
                return css``;
            case 'Error':
                return css``;
            case 'Warning':
                return css``;
            case 'ErrorMessage':
                return css``;
            case 'SuccessMessage':
                return css``;
            case 'ValidationError':
                return css``;
        }
    })(),
});
