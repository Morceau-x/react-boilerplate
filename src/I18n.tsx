import i18n, { InitOptions } from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import HttpApi from 'i18next-http-backend';
import { initReactI18next } from 'react-i18next';
import { Logger } from './Logger';

export const localesNames = {
	en: 'English',
	fr: 'Fran√ßais (French)',
};

export const supportedLocales = Object.keys(localesNames) as (keyof typeof localesNames)[];

export const i18nOptions: InitOptions = {
	ns: [],
	fallbackLng: 'en',
	supportedLngs: supportedLocales,
	nonExplicitSupportedLngs: true,
	load: 'currentOnly',
	lowerCaseLng: true,
	cleanCode: true,
	debug: process.env.NODE_ENV === 'development',
	detection: {
		order: ['localStorage', 'querystring', 'navigator'],
		caches: ['localStorage'],
		lookupQuerystring: 'lng',
		lookupLocalStorage: 'i18nextLng',
	},
	interpolation: {
		escapeValue: false,
	},
	backend: {
		loadPath: '/locales/{{lng}}.json',
	},
	react: {
		useSuspense: false,
		bindI18n: 'loaded languageChanged',
	},
};

i18n.use(initReactI18next)
	.use(LanguageDetector)
	.use(HttpApi)
	.init(i18nOptions)
	.then(Logger.logValueDebug('I18n', 'Initialized'))
	.catch(Logger.errorValueDebug('I18n', 'Initialization error'));

export const i18nInstance = i18n;
