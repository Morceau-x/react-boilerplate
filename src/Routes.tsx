import React, { useMemo } from 'react';
import { generatePath, useRoutes } from 'react-router';
import { useParams, useSearchParams } from 'react-router-dom';
import { Home } from './ui/pages/Home';
import { Game } from './ui/pages/Game';
import { DecoderFunction, field, number, record } from 'typescript-json-decoder';
import { jsonDecode } from './utils/JsonDecoder';
import { concatenatePath, getFragment, PathParams } from './utils/RoutesUtils';
import { NotFound } from './ui/pages/NotFound';
import { History } from './ui/components/CustomBrowserRouter';
import { useAppDispatch } from './redux/ReduxTypes';

export type Paths = {
    '/home': undefined;
    '/game/:width/:height/:bombs': { width: number; height: number; bombs: number };
    '404': undefined;
    '': undefined;
};

type PathsWithParams = {
    [Path in keyof Paths]: PathParams<Path>;
};

type AppRoutesType = {
    [Path in keyof Paths]: Paths[Path] extends undefined
        ? { render: React.ReactNode }
        : {
              decoder: DecoderFunction<Paths[Path]>;
              render: (params: Paths[Path]) => React.ReactNode;
          };
};

const appRoutesRenderers: AppRoutesType = {
    '/home': { render: <Home /> },
    '/game/:width/:height/:bombs': {
        render: (params) => <Game {...params} />,
        decoder: (data) =>
            record({
                width: field('width', number),
                height: field('height', number),
                bombs: field('bombs', number),
            })(data),
    },
    '404': { render: <NotFound /> },
    '': { render: <Home /> },
};

type AppRouteRendererProps = { renderer: typeof appRoutesRenderers[keyof typeof appRoutesRenderers] };

const AppRouteRenderer: React.FC<AppRouteRendererProps> = ({ renderer }) => {
    if ('decoder' in renderer) {
        const params = useParams();
        const searchParams = Object.fromEntries(useSearchParams()[0].entries());
        const fragment = getFragment(History.location.hash);
        const decodedValue = jsonDecode(renderer.decoder, {
            ...fragment,
            ...searchParams,
            ...params,
        });

        switch (decodedValue.status) {
            case 'Ok':
                return <>{renderer.render(decodedValue.data)}</>;
            case 'Err':
                return <NotFound />;
        }
    } else {
        return <>{renderer.render}</>;
    }
};

export const AppRoutes: React.FC = () => {
    useAppDispatch();
    useParams();
    const routesObject = useMemo(
        () =>
            Object.entries(appRoutesRenderers).map(([key, renderer]) => ({
                path: key,
                element: <AppRouteRenderer renderer={renderer} />,
            })),
        []
    );
    return useRoutes(routesObject);
};

export const generateAppPath = <K extends keyof Paths>(
    path: K,
    params: PathsWithParams[K],
    query?: Partial<Paths[K]>,
    fragment?: Partial<Paths[K]>
): string => {
    return concatenatePath(generatePath(path, params), query, fragment);
};
