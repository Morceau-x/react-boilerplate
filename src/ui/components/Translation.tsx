import React, { PropsWithChildren } from 'react';
import { I18nextProvider, useTranslation } from 'react-i18next';
import { TranslationLoadingPage } from '../pages/TranslationLoadingPage';
import { i18nInstance } from '../../I18n';

export const Translation: React.FC<PropsWithChildren> = ({ children }) => {
    return (
        <I18nextProvider i18n={i18nInstance}>
            {(() => {
                const language = useTranslation();
                return language.ready ? children : <TranslationLoadingPage />;
            })()}
        </I18nextProvider>
    );
};
