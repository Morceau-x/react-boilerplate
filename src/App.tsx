import React, { useEffect } from 'react';
import { Provider } from 'react-redux';
import { store } from './Store';
import { createRoot } from 'react-dom/client';
import { Logger } from './Logger';
import { useWindowDimensionsChange } from './hooks/useResize';
import { Translation } from './ui/components/Translation';
import { injectGlobal } from '@emotion/css';
import { Fonts } from './assets/fonts/Fonts';
import { AppRoutes } from './Routes';
import { CustomBrowserRouter } from './ui/components/CustomBrowserRouter';
import { useAppDispatch } from './redux/ReduxTypes';
import { TestSlice } from './redux/Test';

const App: React.FC = () => {
    useWindowDimensionsChange();
    const dispatch = useAppDispatch();

    useEffect(() => {
        dispatch(TestSlice.actions.actionUsedBySaga({ value: 'example' }));
    }, []);

    return (
        <Translation>
            <CustomBrowserRouter>
                <AppRoutes />
            </CustomBrowserRouter>
        </Translation>
    );
};

const rootElement = document.getElementById('app');

if (rootElement) {
    createRoot(rootElement, {
        onRecoverableError: Logger.warnValueDebug('App'),
        identifierPrefix: 'react',
    }).render(
        <Provider store={store}>
            <App />
        </Provider>
    );
}

injectGlobal`
	@font-face {
		font-family: 'Roboto';
		font-style: normal;
		font-weight: 400;
		font-display: swap;
		src: local('Roboto'), local('Roboto-Regular'), url(${Fonts.Roboto.Regular}) format('truetype');;
	}

	@font-face {
		font-family: 'Roboto';
		font-style: normal;
		font-weight: 700;
		font-display: swap;
		src: local('Roboto Bold'), local('Roboto-Bold'), url(${Fonts.Roboto.Bold}) format('truetype');;
	}

	body {
		font-family: 'Roboto', 'Source Sans Pro', 'Trebuchet MS', 'Lucida Grande', 'Bitstream Vera Sans', 'Helvetica Neue', sans-serif;
		margin: 0;
		padding: 0;
		background-color: #fafafa;
		color: rgba(0, 0, 0, 0.87);
		overflow: hidden;
	}
`;

export const test =
    process.env.NODE_ENV === 'test'
        ? {
              rootElement,
              App,
          }
        : {};
