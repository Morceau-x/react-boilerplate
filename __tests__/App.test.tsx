const divElement = document.createElement('div');
const spy = jest.spyOn(document, 'getElementById').mockReturnValue(divElement);
process.env.NODE_ENV = 'test';

import * as React from 'react';
import { test } from '../src/App';
import { create } from 'react-test-renderer';
import * as Translation from '../src/ui/components/Translation';
import * as CustomBrowserRouter from '../src/ui/components/CustomBrowserRouter';
import * as AppRoutes from '../src/Routes';

jest.mock('react-redux', () => ({ Provider: jest.fn() }));
jest.mock('react-dom/client', () => ({ createRoot: () => ({ render: jest.fn() }) }));
jest.mock('@emotion/css', () => ({ injectGlobal: jest.fn() }));
jest.mock('../src/Store', () => ({}));
jest.mock('../src/Logger', () => ({ Logger: { warnValueDebug: jest.fn() } }));
jest.mock('../src/hooks/useResize', () => ({ useWindowDimensionsChange: jest.fn() }));
jest.mock('../src/ui/components/Translation', () => ({ Translation: jest.fn() }));
jest.mock('../src/assets/fonts/Fonts', () => ({ Fonts: { Roboto: { Regular: 'roboto-regular' } } }));
jest.mock('../src/Routes', () => ({ AppRoutes: jest.fn() }));
jest.mock('../src/ui/components/CustomBrowserRouter', () => ({ CustomBrowserRouter: jest.fn() }));

describe('App', () => {
    it('rootElement', () => {
        expect(test.rootElement).toEqual(divElement);
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith('app');
    });
    it('render', () => {
        jest.spyOn(Translation, 'Translation').mockImplementation(({ children }) => <div className='Translation'>{children}</div>);
        jest.spyOn(CustomBrowserRouter, 'CustomBrowserRouter').mockImplementation(({ children }) => (
            <div className='CustomBrowserRouter'>{children}</div>
        ));
        jest.spyOn(AppRoutes, 'AppRoutes').mockImplementation(() => <div className='AppRoutes'>Router content</div>);

        const elem = test.App ? test.App({}) : undefined;
        const tree = elem ? create(elem).toJSON() : '';
        expect(tree).toMatchSnapshot();
    });
});
