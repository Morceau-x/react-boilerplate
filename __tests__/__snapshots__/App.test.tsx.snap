// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render 1`] = `
<div
  className="Translation"
>
  <div
    className="CustomBrowserRouter"
  >
    <div
      className="AppRoutes"
    >
      Router content
    </div>
  </div>
</div>
`;
